include/drivers/rtio.h

```diff
@@ -0,0 +1,482 @@
+			  sizeof(struct rtio_mempool_block) + _maxsz,	\
+			  _nmax,					\
+			  4);						\
+									\
+	struct rtio_mempool_block_allocator _name = {			\
+	.allocator = {							\
+	.alloc = rtio_mempool_block_alloc,				\
+	.free = rtio_mempool_block_free					\
+},									\
+	.mempool = &rtio_pool_mempool_##_name				\
+}
+
+
+/**
+ * @brief Allocate a rtio_block from an rtio_pool with a timeout
+ *
+ * This call is not safe to do with a timeout other than K_NO_WAIT
+ * in an interrupt handler.
+ *
+ * The allocator *must* have a memory layout equivalent to
+ * struct rtio_block_allocator
+ */
+static inline int rtio_block_alloc(void *allocatorv,
+				   struct rtio_block **block,
+				   size_t size,
+				   u32_t timeout)
+{
+	struct rtio_block_allocator *allocator =
+		(struct rtio_block_allocator *)allocatorv;
+
+	return allocator->alloc(allocator, block, size, timeout);
+
+}
+
+/**
+ * @brief Free a rtio_block back to the allocator
+ */
+static inline void rtio_block_free(void *allocatorv, struct rtio_block *block)
+{
+	struct rtio_block_allocator *allocator =
+		(struct rtio_block_allocator *)allocatorv;
+	allocator->free(allocator, block);
+}
+
+/**
+ * @brief Output config describing where and when to push a blocks
+ *
+ * Output config should at least have a valid allocator and k_fifo pointer given
+ *
+ * Drivers should make a best attempt at fulfilling the policy of when
+ * to return and notify the caller the block is ready by pushing it
+ * into the rtio_pipe.
+ *
+ * If the buffer is full before the time or the desired number of
+ * items is fulfilled its likely an error on the authors part and should
+ * be logged as a warning when possible. This is a buffer overrun
+ * scenario on the reader side.
+ *
+ * The behavior when the time is K_FOREVER and number is 0 the buffer
+ * will be marked completed immediately and no read or write will take place.
+ */
+struct rtio_output_config {
+	/**
+	 * Allocator for rtio_blocks
+	 */
+	struct rtio_allocator *allocator;
+
+	/**
+	 * Fifo to output blocks to
+	 */
+	struct k_fifo *fifo;
+
+	/** Time in cycles to read before making the block ready
+	 *
+	 * Should be set using something like K_MSEC(5) or K_FOREVER
+	 */
+	u32_t time;
+
+	/** The number of bytes to read before making the block ready
+	 * Note that the byte size should be computed by the driver from
+	 * the size of the sample set multiplied by some number of samples
+	 */
+	size_t byte_size;
+};
+
+/**
+ * @brief Common trigger sources
+ */
+enum rtio_trigger_source {
+GPIO_IRQ,
+COUNTER_IRQ,
+FUNCTION_CALL,
+
+/* All driver specific sources should use this plus an additive value
+ * to label the trigger source
+ */
+RTIO_TRIGGER_SOURCE_COUNT
+};
+
+/**
+ * @brief Trigger source configuration
+ *
+ * The source of the trigger may be an internal event along with the
+ * associated GPIO device and pin, a counter timeout, or a function call
+ */
+struct rtio_trigger_config {
+	enum rtio_trigger_source source;
```

Comment:
How the configuration for sensor internal trigger could be added to this struct?

Reply:
In the case of a sensor providing triggers like data ready, fifo watermark met, etc each device
would need to provide its own struct which is then assigned to the void* device_config which is part of rtio_config

Reply:
Then you will have the same issue like here: https://github.com/zephyrproject-rtos/zephyr/pull/17434#discussion_r320730133: The API will be error prone.

Reply:
So possibilities to more immediately resolve this issue are by allowing device drivers to extend the rtio_configuration struct by embedding it as the first member. Its still effectively just a pointer to memory being passed around, but a container_of like macro could be used then, or even just casting would be fine.

In a larger change to this that @mbolivar and I discussed the other day, making the device configurations statically defined in dts and providing an id to the configuration you'd like to use here instead could avoid a few issues. The issue you bring up with the void*, issues @andrewboie brought up about this api being not user-mode safe, and perhaps allowing for some optimization at the driver level.

Reply:
I'm not seeing enough context to understand how or why device configurations would be expressed in devicetree.  Without looking at what this does, but from first principles of a streaming infrastructure, I would expect an RTIO instance to connect somehow to a device that is capable of feeding data.  It's on that device that there would be device-specific API to configure whatever generates an event that indicates that data is ready.

Depending on responsiveness requirements, that event might use `k_poll_signal()` (if RTIO is animated by an external thread) or `k_sem_give()` (if RTIO has a dedicated thread) to indicate data readiness.  Rather like @pizi-nordic has for `k_work_triggered()`[sic] in #14041.  So in short a device that's designed to be an RTIO source should be "pollable" in some sense.

But I'll have to look more closely to be able to make more coherent noises.  I'll add it to the queue.

Reply:
The device stream is pollable by directly polling on the output FIFO.